src/site/content/en/blog/kapwing/index.md

```diff
@@ -0,0 +1,415 @@
+  },
+});
+```
+
+We pass a version and define an upgrade function. This is used for
+initialization or to update our schema when necessary.
+
+We also pass error handling callbacks blocked and blocking, which we have
+found useful in preventing issues for users with unstable systems. Finally,
+note our definition of a primary key `keyPath`. In our case this is a unique
+ID we call `mediaLibraryID`. When a user adds a piece of media to our system,
+whether via our uploader or a third party extension, we add the media to our
+media library with the code below:
+
+```js
+export async function addAsset(mediaLibraryID: string, file: File) {
+  return runWithAssetMutex(mediaLibraryID, async () => {
+    const assetAlreadyInStore = await (await openIdb).get(
+      "assets",
+      mediaLibraryID
+    );    
+    if (assetAlreadyInStore) return;
+        
+    const idbVideo: IdbVideo = {
+      file,
+      mediaLibraryID,
+    };
+
+    await (await openIdb).add("assets", idbVideo);
+  });
+```
+
+`runWithAssetMutex` is our own internally defined function that serializes
+indexedDB access. This is required for any read-modify-write type operations,
+as the IndexedDB API is asynchronous. 
+
+Now let's take a look at how we access files. Below is our `getAsset` function:
+
+```js
+export async function getAsset(
+  mediaLibraryID: string,
+  source: LayerSource | null | undefined,
+  location: string
+): Promise<IdbAsset | undefined> {
+  let asset: IdbAsset | undefined;
+  const { idbCache } = window;
+  const assetInCache = idbCache[mediaLibraryID];
+  if (assetInCache && assetInCache.status === "complete") {
+    asset = assetInCache.asset;
+  } else if (assetInCache && assetInCache.status === "pending") {
+    asset = await new Promise((res) => {
+      assetInCache.subscribers.push(res);
+    }); 
+  } else {
+    idbCache[mediaLibraryID] = { subscribers: [], status: "pending" };
+    asset = (await openIdb).get("assets", mediaLibraryID);
+    idbCache[mediaLibraryID].asset = asset;
+    idbCache[mediaLibraryID].subscribers.forEach((res: any) => {
+      res(asset);
+    });
+    delete (idbCache[mediaLibraryID] as any).subscribers;
+    if (asset) idbCache[mediaLibraryID].status = "complete";
+    else idbCache[mediaLibraryID].status = "failed";
+  } 
+  return asset;
+} 
+```
+
+We have our own data structure, `idbCache`, that is used to minimize indexedDB
+accesses. While indexedDB is fast, accessing local memory is faster. We
+recommend this approach so long as you manage the size of the cache.
+
+Also note the `subscribers` array, which is used to prevent simultaneous
+accesses to indexedDB, which otherwise would be common on load.
+
+### Web Audio API
+
+Audio visualization is incredibly important for video editing. To understand
+why, take a look at a screenshot from the editor:
+
+{% Img 
+   src="image/VbsHyyQopiec0718rMq2kTE1hke2/6MqSyhNqFl23wN3GWKbP.png", 
+   alt="Kapwing editor has several templates and capabilities",
```

Comment:
This alt text should be more descriptive. As this stands, a person reliant on assistive technologies may not be able to know what's in the image.

Reply:
I've updated this significantly:

> Kapwing's editor has a menu for media, including several templates and custom elements, including some templates which are specific to certain platforms like LinkedIn; a timeline which separates video, audio, and animation; canvas editor with export quality options; a preview of the video; and more capabilities.

There's a lot going on, and I don't know that every piece needs to be written out as it's not all useful in the context of this blog post. However, the new copy has significantly more value.